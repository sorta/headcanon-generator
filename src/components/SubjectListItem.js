import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { allowEdit } from '../utils'

class SubjectListItem extends Component {
  static propTypes = {
    subject: PropTypes.shape({
      name: PropTypes.string,
      fandomKey: PropTypes.string,
    }),
    updateSubject: PropTypes.func,
    deleteSubject: PropTypes.func,
    index: PropTypes.string,
    setAvailability: PropTypes.func,
    subjectAvailability: PropTypes.object,
    fandomAvailability: PropTypes.object,
    isManaging: PropTypes.func,
  };
  static defaultProps = {
    subject: { name: '', fandomKey: '' },
    updateSubject: () => {},
    deleteSubject: () => {},
    index: '',
    setAvailability: () => {},
    subjectAvailability: {},
    fandomAvailability: {},
    isManaging: () => false,
  };

  handleChange = (event) => {
    const updatedSubject = {
      ...this.props.subject,
      [event.currentTarget.name]: event.currentTarget.value,
    };

    this.props.updateSubject(this.props.index, updatedSubject);
  };

  toggleAvailability = (event) => {
    this.props.setAvailability('subject', this.props.index, event.currentTarget.checked);
  };

  render() {
    let availCheckbox = null;
    if (!({}.hasOwnProperty.call(this.props.fandomAvailability, this.props.subject.fandomKey))) {
      availCheckbox = (
        <input
          id={`isAvailable_subj_${this.props.index}`}
          name="isAvailable"
          type="checkbox"
          className="cb-fake"
          checked={!({}.hasOwnProperty.call(this.props.subjectAvailability, this.props.index))}
          onChange={this.toggleAvailability}
        />
      );
    }

    if (!this.props.isManaging() || !allowEdit) {
      return (
        <li key={this.props.index} className="SubjectListItem list-item control-row">
          {availCheckbox}
          <label
            htmlFor={`isAvailable_subj_${this.props.index}`}
            className={availCheckbox === null ? "SubjectListItem-text strike" : "SubjectListItem-text"}
          >
            {this.props.subject.name}
          </label>
        </li>
      );
    }

    return (
      <li key={this.props.index} className="SubjectListItem list-item control-row">
        {availCheckbox}
        <label htmlFor={`isAvailable_subj_${this.props.index}`}></label>
        <input name="name" type="text" placeholder="Subject Name" className="control-stretch"
          value={this.props.subject.name} onChange={this.handleChange}
        />
        <button
          className="SubjectListItem-delete"
          onClick={() => this.props.deleteSubject(this.props.index)}
        >-</button>
      </li>
    );
  }
}

export default SubjectListItem;
