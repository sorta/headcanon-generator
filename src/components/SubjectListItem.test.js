import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import SubjectListItem from './SubjectListItem';

const props = {
  subject: { name: 'Test 1', fandomKey: 'f_test_1' },
  subjectAvailability: {},
  updateSubject: jest.fn(),
  deleteSubject: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
  index: 's_test_1',
  fandomAvailability: {},
};

// Tests for regular mode
describe("SubjectListItem", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <SubjectListItem { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a SubjectListItem li element', () => {
    const component = shallow(<SubjectListItem { ...props } />);
    expect(component.find('li.SubjectListItem.list-item').exists()).toBe(true);
  });

  it('has a checkbox and label', () => {
    const component = shallow(<SubjectListItem { ...props } />);
    expect(component.find(`input[type="checkbox"][name="isAvailable"]#isAvailable_subj_${props.index}`).exists()).toBe(true);
    expect(component.find(`label[htmlFor="isAvailable_subj_${props.index}"]`).exists()).toBe(true);
  });

  it('clicking checkbox calls setAvailability', () => {
    const component = mount(<SubjectListItem { ...props } />);
    component.find('input[type="checkbox"][name="isAvailable"]').simulate('change', { target: { checked: false } });
    expect(props.setAvailability.mock.calls.length).toEqual(1);
  });

  it('does not have edit elements', () => {
    const component = shallow(<SubjectListItem { ...props } />);
    expect(component.find('input[name="name"]').exists()).toBe(false);
    expect(component.find('button.SubjectListItem-delete').exists()).toBe(false);
  });
});

// Tests for managing mode
const mProps = { ...props };
mProps.isManaging = () => true;
describe("SubjectListItem in managing mode", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <SubjectListItem { ...mProps } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a SubjectListItem li element', () => {
    const component = shallow(<SubjectListItem { ...mProps } />);
    expect(component.find('li.SubjectListItem.list-item').exists()).toBe(true);
  });

  it('has a checkbox and label', () => {
    const component = shallow(<SubjectListItem { ...mProps } />);
    expect(component.find(`input[type="checkbox"][name="isAvailable"]#isAvailable_subj_${mProps.index}`).exists()).toBe(true);
    expect(component.find(`label[htmlFor="isAvailable_subj_${mProps.index}"]`).exists()).toBe(true);
  });

  it('clicking checkbox calls setAvailability', () => {
    const component = mount(<SubjectListItem { ...mProps } />);
    component.find('input[type="checkbox"][name="isAvailable"]').simulate('change', { target: { checked: false } });
    expect(mProps.setAvailability.mock.calls.length).toEqual(2);
  });

  it('has edit elements', () => {
    const component = shallow(<SubjectListItem { ...mProps } />);
    expect(component.find('input[name="name"]').exists()).toBe(true);
    expect(component.find('button.SubjectListItem-delete').exists()).toBe(true);
  });

  it('clicking delete button calls deleteSubject', () => {
    const component = shallow(<SubjectListItem { ...mProps } />);
    component.find('button.SubjectListItem-delete').simulate('click');
    expect(mProps.deleteSubject.mock.calls.length).toEqual(1);
  });

  it('changing name value calls updateSubject', () => {
    const component = mount(<SubjectListItem { ...mProps } />);
    component.find('input[name="name"]').simulate('change', { target: { value: 'TEST' } });
    expect(mProps.updateSubject.mock.calls.length).toEqual(1);
  });
});
