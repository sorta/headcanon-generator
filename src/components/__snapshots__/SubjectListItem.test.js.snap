// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectListItem in managing mode renders and matches our snapshot 1`] = `
<li
  className="SubjectListItem list-item control-row"
>
  <input
    checked={true}
    className="cb-fake"
    id="isAvailable_subj_s_test_1"
    name="isAvailable"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="isAvailable_subj_s_test_1"
  />
  <input
    className="control-stretch"
    name="name"
    onChange={[Function]}
    placeholder="Subject Name"
    type="text"
    value="Test 1"
  />
  <button
    className="SubjectListItem-delete"
    onClick={[Function]}
  >
    -
  </button>
</li>
`;

exports[`SubjectListItem renders and matches our snapshot 1`] = `
<li
  className="SubjectListItem list-item control-row"
>
  <input
    checked={true}
    className="cb-fake"
    id="isAvailable_subj_s_test_1"
    name="isAvailable"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="SubjectListItem-text"
    htmlFor="isAvailable_subj_s_test_1"
  >
    Test 1
  </label>
</li>
`;
